l0_sampling/sketch.cpp

```diff
@@ -1,52 +1,68 @@
+  const unsigned long long int num_guesses = guess_gen(n);
+  buckets = std::vector<Bucket_Boruvka>(num_buckets * num_guesses);
 }
 
 void Sketch::update(Update update ) {
-  for (unsigned j = 0; j < buckets.size(); j++){
-    if (buckets[j].contains(update.index+1)){
-      buckets[j].a += update.delta;
-      buckets[j].b += update.delta*(update.index+1); // deals with updates whose indices are 0
-      buckets[j].c += (update.delta*PrimeGenerator::power(buckets[j].r,update.index+1,random_prime))%random_prime;
-      buckets[j].c = (buckets[j].c + random_prime)%random_prime;
+  const unsigned long long int num_buckets = bucket_gen(n);
+  const unsigned long long int num_guesses = guess_gen(n);
+  for (unsigned i = 0; i < num_buckets; ++i) {
+    for (unsigned j = 0; j < num_guesses; ++j) {
+      unsigned bucket_id = i * num_guesses + j;
+      XXH64_hash_t bucket_seed = XXH64(&bucket_id, sizeof(bucket_id), seed);
+      int128_t r = 2 +  bucket_seed % (large_prime - 3);
+      if (buckets[bucket_id].contains(update.index+1, bucket_seed, 1 << j)){
+        buckets[bucket_id].a += update.delta;
+        buckets[bucket_id].b += update.delta*(update.index+1); // deals with updates whose indices are 0
+        buckets[bucket_id].c = static_cast<uint128_t>(
+              (static_cast<int128_t>(buckets[bucket_id].c)
+              + static_cast<int128_t>(large_prime)
+              + (update.delta*PrimeGenerator::power(r,(uint128_t) update
+              .index+1, large_prime) % static_cast<int128_t>(large_prime)))
+              % large_prime
+              );
+      }
     }
   }
 }
 
-Update Sketch::query(){
-  if (already_quered){
+Update Sketch::query() {
+  if (already_quered) {
     throw MultipleQueryException();
   }
   already_quered = true;
   bool all_buckets_zero = true;
-  for (unsigned i = 0; i < buckets.size(); i++){
-    Bucket& b = buckets[i];
-    if (b.a != 0 || b.b != 0 || b.c != 0) {
-      all_buckets_zero = false;
-    }
-    if (b.a != 0 && b.b % b.a == 0 && b.b/b.a > 0 && b.b/b.a <= n && b.contains(b.b/b.a)
-      && (b.c - b.a*PrimeGenerator::power(b.r,b.b/b.a,random_prime))% random_prime == 0) {
-      //cout << "Passed all tests: " << "b.a: " << b.a << " b.b: " << b.b << " b.c: " << b.c << " Guess: " << b.guess_nonzero << " r: " << b.r << endl;
-      //cout << "String: " << b.stuff << endl;
-      return {b.b/b.a - 1,b.a}; // 0-index adjustment
+  const unsigned long long int num_buckets = bucket_gen(n);
+  const unsigned long long int num_guesses = guess_gen(n);
+  for (unsigned i = 0; i < num_buckets; ++i) {
+    for (unsigned j = 0; j < num_guesses; ++j) {
+      Bucket_Boruvka& b = buckets[i*num_guesses+j];
+      if (b.a != 0 || b.b != 0 || b.c != 0) {
+        all_buckets_zero = false;
+      }
+      unsigned bucket_id = i * num_guesses + j;
+      XXH64_hash_t bucket_seed = XXH64(&bucket_id, sizeof(bucket_id), seed);
+      uint128_t r = 2 + bucket_seed % (large_prime - 3);
```

Comment:
I get that r should be {-1,0,1}, but why we do we have (large prime - 3).

Reply:
r should not be {-1,0,1}, and since we are working with mod large_prime, r should not be {0,1,large_prime-1}.
This means we want an integer in the range [2,large_prime-2].

rand%m gives an integer in the range [0,m-1].
bucket_seed % (large_prime - 3) gives an integer in the range [0,large_prime - 4].
2 + bucket_seed % (large_prime - 3) gives an integer in the range [2,large_prime - 2], which is what we're looking for.